syntax = "proto3";

package grpc;

message Empty {
}

message Worker {
    int64 id = 1;
}

enum TaskType {
    idle = 0;
    map = 1;
    reduce = 2;
    shut_down = 3;
}

message Task {
    TaskType task_type = 1;
    int64 id = 2;
    int64 num_red_tasks = 3;
    repeated string data = 4;
}

message WordList {
    repeated string word_list = 1;
}

message MapResults {
    map<string, WordList> map_results = 1;
}

message ReduceResults {
    map<string, int32> reduce_results = 1;
    int32 bucket_id = 2;
}

service MapReduce {
    rpc get_worker_task (Worker) returns (Task);
    rpc finish_map_task (MapResults) returns (MapResults);
    rpc finish_reduce_task (ReduceResults) returns (Task);
    rpc worker_down(Worker) returns (Empty);
} 